src/sos_filter.c

```diff
@@ -0,0 +1,149 @@
+        sf->filter_state[section][0] =
+            fixed_mul(sf->filter[section][1], cur_val)
+            - fixed_mul(sf->filter[section][3], next_val)
+            + (sf->filter_state[section][1]);
+        sf->filter_state[section][1] =
+            fixed_mul(sf->filter[section][2], cur_val)
+            - fixed_mul(sf->filter[section][4], next_val);
+        cur_val = next_val;
+    }
+
+    return cur_val;
+}
+
+
+// resets the filter state (zi) to 0 to start a new filtering session
+void
+reset_filter_state(struct sos_filter *sf) {
+    for (uint8_t i = 0; i < MAX_SECTIONS; i++){
+        sf->filter_state[i][0] = 0;
+        sf->filter_state[i][1] = 0;
+    }
+}
+
+// mark filter as not ready for use, ready to be updated
+void
+reset_filter(struct sos_filter *sf, uint8_t n_sections) {
+    if (n_sections > MAX_SECTIONS) {
+        shutdown("sos_filter: too many filter sections");
+    }
+    sf->n_sections = n_sections;
+    sf->last_section_idx = -1;
+    sf->is_filter_ready = 0;
+}
+
+// Create an sos_filter
+void
+command_config_sos_filter(uint32_t *args)
+{
+    struct sos_filter *sf = oid_alloc(args[0]
+                            , command_config_sos_filter, sizeof(*sf));
+    reset_filter(sf, 0);
+}
+DECL_COMMAND(command_config_sos_filter, "config_sos_filter oid=%c");
+
+// Lookup an sos_filter
+struct sos_filter *
+sos_filter_oid_lookup(uint8_t oid)
+{
+    return oid_lookup(oid, command_config_sos_filter);
+}
+
+// Set one section of the filter
+void
+command_config_filter_section(uint32_t *args)
+{
+    struct sos_filter *sf = sos_filter_oid_lookup(args[0]);
+    uint8_t n_sections = args[1];
+    uint8_t section_idx = args[2];
+    // if this is the first filter section, initalize
+    if (section_idx <= 0)  {
+        reset_filter(sf, n_sections);
+    } else {
+        // validate that this new section is in order
+        if (n_sections != sf->n_sections) {
+            shutdown("Filter size mismatch");
+        }
+        if (section_idx != sf->last_section_idx + 1) {
+            shutdown("Filter section out of order");
+        }
+    }
+
+    // if the number of configured sections positive
+    if (n_sections > 0) {
+        // copy section data
+        const uint8_t arg_base = 3;
+        for (uint8_t i = 0; i < SECTION_WIDTH; i++) {
+            sf->filter[section_idx][i] = args[i + arg_base];
+        }
+        sf->last_section_idx = section_idx;
+        /*
+        // DEBUG: uncomment to log filter section contents
+        for (uint8_t i = 0; i < SECTION_WIDTH; i++) {
+            output("section[%c][%c]=%i", section_idx, i,
+                                (int32_t)sf->sos[section_idx][i]);
+        }*/
+    }
+
+    // check for last section
+    if (sf->last_section_idx == sf->n_sections - 1) {
+        sf->is_filter_ready = 1;
+    }
+}
+DECL_COMMAND(command_config_filter_section
+    , "config_sos_filter_section oid=%c n_sections=%c"
+    " section_idx=%c sos0=%i sos1=%i sos2=%i sos3=%i sos4=%i");
```

Comment:
By convention, any mcu command starting with "config_" will create an oid.  So, best to rename this command.